### DISCUSSION TITLE
I can't add a kernel to Jupyter Lite

### DISCUSSION THREAD

**Discussion by twiine** [jupyterlite/jupyterlite#1136] [Q&A] [answered]
Hello,

I installed jupyterlite-core and jupyterlite-pyodide-kernel via pip. Then I create my jupyter lite dist with:

jupyter lite build --output-dir=dist

To access it I use apache (copied dist to /var/www/html). 

I can open jupyter lite but I cannot select any kernels. I also tried to use a local copy of pyodide: (added --pyodide=pyodide-0.23.4.tar.bz2) and I installed the echo kernel. Both didn't help.

I think I am missing something in the installation. Thank you for any hints.

Versions: jupyter-core-5.3.1 jupyterlite-core-0.1.2 jupyterlite-pyodide-kernel-0.1.1

Here is the output of the build:

static:jupyter-lite.json
.  pre_status:static:jupyter-lite.json
    tarball:         jupyterlite-app-0.1.2.tgz 7MB
    output:          /home/ubuntu/lite/dist
    lite dir:        /home/ubuntu/lite
    apps:
    sourcemaps:      True
    unused packages: True
archive:archive
contents:contents
jupyterlite-pyodide-kernel-pyodide:pyodide
lite:jupyter-lite.json
mathjax:status
mimetypes:jupyter-lite.json
serve:contents
settings:overrides
translation:translation
.  status:archive:archive
[LiteBuildApp] No archive (yet): lite-jupyterlite.tgz
.  status:contents:contents
    contents: 0 files
.  status:jupyterlite-pyodide-kernel-pyodide:pyodide
     URL: None
 archive: []
   cache: 477 files
   local: 0 files
.  status:lite:jupyter-lite.json
[LiteBuildApp]     jupyter-lite.(json|ipynb): 0 files
.  status:mathjax:status
     jupyter-server-mathjax: None
     configured MathJax dir: /home/ubuntu/lite/traitlets.Undefined
     effective MathJax dir: None
.  status:mimetypes:jupyter-lite.json
    filetypes:         26
.  status:serve:contents
    url: http://127.0.0.1:8000/
    server: stdlib
    headers:
.  status:settings:overrides
    overrides.json: 0
.  status:translation:translation
    translation files: 0 files
static:output_dir
.  pre_init:static:output_dir
static:unpack
.  init:static:unpack
/home/ubuntu/.local/lib/python3.10/site-packages/jupyterlite_core/optional.py:22: UserWarning: install libarchive-c for better perfomance when working with archives: No module named 'libarchive'
  warnings.warn(hint.format(error=error))
/home/ubuntu/.local/lib/python3.10/site-packages/jupyterlite_core/optional.py:22: UserWarning: [lite] [translation] install `jupyterlab_server` to load translations: No module named 'jupyterlab_server'
  warnings.warn(hint.format(error=error))
translation:copy
.  build:translation:copy
federated_extensions:patch
mimetypes:patch
.  post_build:federated_extensions:patch
.  post_build:mimetypes:patch

**Comment by jtpio** [comment]
@twiine would you be able to check if the pyodide extension is correctly installed and detected?

```
pip install jupyterlab~=3.5.0
jupyter labextension list
```

Should output something like the following:

```
@jupyterlite/pyodide-kernel-extension v0.1.1 enabled  X (python, jupyterlite_pyodide_kernel)
```

**Reply by twiine** [reply]
Output if I run jupyter labextension list

> /home/ubuntu/.local/share/jupyter/labextensions
>       jupyterlab_pygments v0.2.2 enabled OK (python, jupyterlab_pygments)
>       @jupyterlite/pyodide-kernel-extension v0.1.1 enabled OK (python, jupyterlite_pyodide_kernel)

Do I need to install jupyterlab to build jupyter lite? Anyhow, I tried it again but I still don't have a kernel available.

**Reply by jtpio** [reply]
> Do I need to install jupyterlab to build jupyter lite?

It's not needed, but it provides the `jupyter labextension list` command which in this case was useful to double check the extension was installed.

**Reply by jtpio** [reply]
Maybe worth trying to delete the cache with `rm -rf .jupyterlite.doit.db dist` just to make sure, and then re-run `jupyter lite build`.

**Reply by twiine** [reply]
I tried it, but it didn't change anything. 

Is it possible to see which kernels are registerd in the build folder (dist)? I wonder why I don't even find the echo kernel (I greped through the files).

**Reply by jtpio** [reply]
Normally they should end up under `dist/extensions`.

**Reply by twiine** [reply]
There is no extensions folder or file in dist.

**Reply by twiine** [reply, marked as answer]
Hello,

I think I found the issue: 

In federated_extensions.py the path to the labextensions is hard coded to: /usr/share/jupyter/labextensions
As I installed jupyter lite only in the user directory, this path doesn't exist. I replaced the correct path and now it works.

```python
#ENV_EXTENSIONS = Path(sys.prefix) / SHARE_LABEXTENSIONS
ENV_EXTENSIONS = Path('/home/ubuntu/.local/share/jupyter/labextensions/')
``` 

The `jupyter labextension list` was misleading as it uses the codebase from jupyterlab, which apparently has the correct path.

**Reply by JackYinpei** [reply]
Hello，

Hi, I'm wondering where the federated_extensions.py file you mentioned in your reply is located. 

I encountered the same issue when adding pyodide.

Thanks in advance

**Reply by twiine** [reply]
It is located in: ~/.local/lib/python3.10/site-packages/jupyterlite_core/addons/federated_extensions.py

**Reply by JackYinpei** [reply]
Thank you very much for your reply. 

I found this file in another location of mine and made the modifications, but when I ran "jupyter lite build" followed by "serve", there was still no pyodide kernel available to use in the browser. 

I wanted to ask if you encountered any other errors after modifying the federated_extensions.py file, or if you have any other ways to add pyodide to the project.

I've been trying to deploy this project on my machine for the past few days, but it's been taking up a lot of my time. I would really appreciate any help you can provide.

**Reply by twiine** [reply]
Hello,

I installed jupyterlite and the kernel:

pip3 install jupyterlite-core
pip3 install jupyterlite-pyodide-kernel

Then I modified the file as described above and I build it

jupyter lite build   --output-dir dist

That worked for me. BTW I use apache, but that shouldn't matter.

**Reply by augusto-herrmann** [reply]
Hi! Thanks for sharing this workaround.

However, the current version of `federated_extensions.py` does not have this line. There is [a similarly named function](https://github.com/jupyterlite/jupyterlite/blob/c865a593745ef4f8227e41b51043af5317f0e803/py/jupyterlite-core/jupyterlite_core/addons/federated_extensions.py#L36) though, but I couldn't figure out so far how to make it detect the installed `jupyterlite-core` and `jupyterlite-pyodide-kernel` packages. The packages have been installed in user space, with `pip install --user`.

**Reply by martinRenou** [reply]
Just to clarify, the labextension path is not *hard coded*, it's where your Python's `sys.prefix` lives. `python -c "import sys; print(sys.prefix)"` would show you the path of the sys prefix, you can then append "share/jupyter/labextensions/" to what is printed.

**Reply by gautamsarawagi** [reply]
@augusto-herrmann Were you able to figure out the solution? I guess I am also facing the same issue.

**Comment by gautamsarawagi** [comment]
I have tried to built it in the python virtual environment. The order that I have followed: 

1. python -m pip install jupyterlite-core
2. Then jupyter lite init.
3. jupyter lite build --output-dir dist
but getting this error: 

static:jupyter-lite.json
.  pre_status:static:jupyter-lite.json
    tarball:         jupyterlite-app-0.2.3.tgz 9MB
    output:          /home/gautam/Downloads/jupyterlite-standalone/dist
    lite dir:        /home/gautam/Downloads/jupyterlite-standalone
    apps:            
    sourcemaps:      True
    unused packages: True
archive:archive
contents:contents
icons:icons
jupyterlite-pyodide-kernel-pyodide:pyodide
lite:jupyter-lite.json
mimetypes:jupyter-lite.json
serve:contents
settings:overrides
translation:translation
.  status:archive:archive
[LiteBuildApp] No archive (yet): jupyterlite-standalone-jupyterlite.tgz
.  status:contents:contents
    contents: 0 files
.  status:icons:icons
    favicon files: 0 files
.  status:jupyterlite-pyodide-kernel-pyodide:pyodide
     URL: None
 archive: []
   cache: 0 files
   local: 0 files
.  status:lite:jupyter-lite.json
[LiteBuildApp]     jupyter-lite.(json|ipynb): 14 files
.  status:mimetypes:jupyter-lite.json
    filetypes:         26 
.  status:serve:contents
    url: http://127.0.0.1:8000/
    server: stdlib
    headers:
.  status:settings:overrides
    overrides.json: 0
.  status:translation:translation
    translation files: 0 files
static:output_dir
-- pre_init:static:output_dir
static:unpack
.  init:static:unpack
TaskError - taskid:init:static:unpack
PythonAction Error
Traceback (most recent call last):
  File "/home/gautam/Downloads/jupyterlite-standalone/lib/python3.11/site-packages/doit/action.py", line 461, in execute
    returned_value = self.py_callable(*self.args, **kwargs)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gautam/Downloads/jupyterlite-standalone/lib/python3.11/site-packages/jupyterlite_core/addons/static.py", line 128, in _unpack_stdlib
    self.extract_one(self.app_archive, tdp)
  File "/home/gautam/Downloads/jupyterlite-standalone/lib/python3.11/site-packages/jupyterlite_core/addons/base.py", line 312, in extract_one
    libarchive.extract_file(str(archive))
    ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'libarchive' has no attribute 'extract_file'

Can u pls help me resolving it. I am not even able to run the basic jupyter-lite site locally. I am following this [document](https://jupyterlite.readthedocs.io/en/stable/quickstart/standalone.html).

**Comment by augusto-herrmann** [comment]
No, @gautamsarawagi. The problem I had was completely different. I never got any error messages in the build process.

My problem was (or rather is) the installed packages besides the Jupyter ones are never accessible to the Pyodide kernel. I get import errors. Not even `plotly`, which is included in JupyterLite's own `requirements.txt` file.

The workaround I found somewhere (sorry, can't find the source again to cite it) is to just install it in the first cell of every notebook:

```
%pip install -q nbformat plotly
```

**Comment by RMattB1** [comment]
For anyone seeing similar symptoms, but with newer versions, check out 
https://github.com/jupyterlite/jupyterlite/pull/1583
